### PR TITLE
fix(csharp): fixed sample code and created CI job to validate

### LINKED ISSUE #259: Implement build, synth, and test checks in CI
### General Information

* [x] :wave: I may be able to implement this feature request

* [ ] :warning: This feature might incur a breaking change 

### Description

We should make sure PRs are checked for 
- site builds successfully
- sample code `cdk synth`'s successfully
- sample code tests pass

### Proposed Solution

Implement GitHub workflows for all 3 checks to be ran on every PR.

### PR DESCRIPTION
**Pull request by ryparker**
## Summary

This PR implements CI workflows that verify changes do not break builds.

- Fix csharp sample code so that it works with `cdk synth`
- GitHub Workflows that synths the csharp sample code
- Update csharp workshop instructions to reflect the fixes I was required to implement

Fixes #259, https://github.com/aws-samples/aws-cdk-intro-workshop/issues/175, https://github.com/aws-samples/aws-cdk-intro-workshop/issues/152

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: workshop/content/40-dotnet/50-table-viewer/400-expose-table.md [hunk 1]
```diff
@@ -23,8 +23,8 @@ namespace CdkWorkshop
 
     public class HitCounter : Construct
     {
-        public IFunction Handler { get; };
-        public Table MyTable { get; };
+        public readonly IFunction Handler { get; };
+        public readonly Table MyTable { get; };
```
Comment (hakanson) [inline workshop/content/40-dotnet/50-table-viewer/400-expose-table.md:27]: Should either be `readonly` or auto getter with `{ get }` - having both causes this compile error:

    The modifier 'readonly' is not valid for this item [CdkWorkshop]csharp(CS0106)

If `{ get }` then remove trailing semicolon or this compile error:

    Semicolon after method or accessor block is not valid [CdkWorkshop]csharp(CS1597)

It should probably just match what is in [HitCounter.cs L16-L17](https://github.com/aws-samples/aws-cdk-intro-workshop/blob/master/code/csharp/pipelines-workshop/src/CdkWorkshop/HitCounter.cs#L16-L17)

```csharp
        public readonly Function Handler;
        public readonly Table MyTable;
```

Comment (ryparker) [reply workshop/content/40-dotnet/50-table-viewer/400-expose-table.md]: Nice catch. Done 👍🏻

File: code/csharp/main-workshop/src/CdkWorkshop/CdkWorkshop.csproj [hunk 2]
```diff
@@ -2,19 +2,21 @@
...
-    <PackageReference Include="Amazon.CDK.AWS.APIGateway" Version="1.17.1-devpreview" />
-    <PackageReference Include="Amazon.CDK.AWS.DynamoDB" Version="1.17.1-devpreview" />
-    <PackageReference Include="Amazon.CDK.AWS.Lambda" Version="1.17.1-devpreview" />
+    <PackageReference Include="Amazon.CDK" Version="1.117.*" />
+    <PackageReference Include="Amazon.CDK.AWS.APIGateway" Version="1.117.*" />
+    <PackageReference Include="Amazon.CDK.AWS.DynamoDB" Version="1.117.*" />
+    <PackageReference Include="Amazon.CDK.AWS.Lambda" Version="1.117.*" />
 
-    <PackageReference Include="Eladb.DynamoTableViewer" Version="3.0.6" />
+    <PackageReference Include="Eladb.DynamoTableViewer" Version="3.1.*" />
```
Comment (hakanson) [inline code/csharp/main-workshop/src/CdkWorkshop/CdkWorkshop.csproj:15]: See my comment on #152 where I had to use `Version="[0.1.*,3)"` by running `dotnet add package Eladb.DynamoTableViewer --version "[0.1.*,3)"` to not get

> error NU1107: Version conflict detected for Amazon.JSII.Runtime.  Install/reference Amazon.JSII.Runtime 1.31.0 directly to project CdkWorkshop to resolve this issue.

Is there a corresponding doc change to run `dotnet add package Amazon.Jsii.Runtime --version "1.31.*"` to get this entry in the `.csproj` file when running the workshop?

File: workshop/content/40-dotnet/70-advanced-topics/100-pipelines/3000-new-pipeline.md [hunk 3]
```diff
@@ -59,12 +59,11 @@ namespace CdkWorkshop
                     SourceArtifact = sourceArtifact,  // Where to get source code to build
                     CloudAssemblyArtifact = cloudAssemblyArtifact,  // Where to place built source
 
-                    InstallCommands = new [] 
-                    {
-                        "npm install -g aws-cdk"
-                    },
-                    BuildCommands = new [] { "dotnet build src" }, // Language-specific build cmd
-                    SynthCommand = "cdk synth"
+                    InstallCommand = string.Join(
```
Comment (hakanson) [inline workshop/content/40-dotnet/70-advanced-topics/100-pipelines/3000-new-pipeline.md:62]: See my comment on #175 about only needing `"npm install -g aws-cdk"`.

Also, that [String.Join Method](https://docs.microsoft.com/en-us/dotnet/api/system.string.join?view=netcore-3.1) signature looks off.  I was looking for the ` && ` separator somewhere like in this code:

```csharp
InstallCommand = string.Join(" && ",
	new string [] {
		"npm install -g aws-cdk",
		"sudo apt-get install -y dotnet-sdk-3.1"
	}
```

File: code/csharp/main-workshop/src/CdkWorkshop/CdkWorkshop.csproj (hunk 2 above)
Comment (ryparker) [reply code/csharp/main-workshop/src/CdkWorkshop/CdkWorkshop.csproj]: I found that using your versioning of `[0.1.*,3)` enabled me to remove the 2 Jsii dependencies so I decided to implement your versioning instead.

There is one place in the workshop guide that instructs the user to add the package. [I updated that instruction here](https://github.com/aws-samples/aws-cdk-intro-workshop/pull/268/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.yml&file-filters%5B%5D=dotfile#diff-b200b09578c1ea2aacde8f31ddf0e21af2d50a3db364d76ca4428f42dc4a2929R12)

File: code/csharp/main-workshop/src/CdkWorkshop/CdkWorkshop.csproj [hunk 4]
```diff
@@ -2,19 +2,18 @@
...
-    <PackageReference Include="Amazon.CDK.AWS.Lambda" Version="1.17.1-devpreview" />
+    <PackageReference Include="Amazon.CDK" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.AWS.APIGateway" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.AWS.DynamoDB" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.AWS.Lambda" Version="1.*" />
 
-    <PackageReference Include="Eladb.DynamoTableViewer" Version="3.0.6" />
+    <PackageReference Include="Eladb.DynamoTableViewer" Version="[0.1.*,3)" />
 
-    <PackageReference Include="Amazon.Jsii.Analyzers" Version="*" PrivateAssets="all" />
```
Comment (ryparker) [inline code/csharp/main-workshop/src/CdkWorkshop/CdkWorkshop.csproj:17]: No longer necessary. Removed

File: code/csharp/main-workshop/src/CdkWorkshop/CdkWorkshop.csproj [hunk 5]
```diff
@@ -2,19 +2,18 @@
...
-    <PackageReference Include="Amazon.CDK.AWS.APIGateway" Version="1.17.1-devpreview" />
-    <PackageReference Include="Amazon.CDK.AWS.DynamoDB" Version="1.17.1-devpreview" />
-    <PackageReference Include="Amazon.CDK.AWS.Lambda" Version="1.17.1-devpreview" />
+    <PackageReference Include="Amazon.CDK" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.AWS.APIGateway" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.AWS.DynamoDB" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.AWS.Lambda" Version="1.*" />
 
-    <PackageReference Include="Eladb.DynamoTableViewer" Version="3.0.6" />
+    <PackageReference Include="Eladb.DynamoTableViewer" Version="[0.1.*,3)" />
```
Comment (ryparker) [inline code/csharp/main-workshop/src/CdkWorkshop/CdkWorkshop.csproj:15]: This version specification was required in order for Eladb.DynamoTableViewer to share compatible peer dependencies with the latest CDK packages.

File: code/csharp/pipelines-workshop/src/CdkWorkshop/CdkWorkshop.csproj [hunk 6]
```diff
@@ -7,18 +7,16 @@
...
+    <PackageReference Include="Amazon.CDK.AWS.DynamoDB" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.AWS.Lambda" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.Pipelines" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.AWS.CodePipeline" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.AWS.CodeCommit" Version="1.*" />
+    <PackageReference Include="Amazon.CDK.AWS.CodePipeline.Actions" Version="1.*" />
 
-    <PackageReference Include="Eladb.DynamoTableViewer" Version="3.1.2" />
-
-    <PackageReference Include="Amazon.Jsii.Analyzers" Version="*" PrivateAssets="all" />
```
Comment (ryparker) [inline code/csharp/pipelines-workshop/src/CdkWorkshop/CdkWorkshop.csproj:21]: same changes as main-workshop

File: code/csharp/pipelines-workshop/src/CdkWorkshop/PipelineStack.cs [hunk 7]
```diff
@@ -44,12 +44,13 @@ public WorkshopPipelineStack(Construct parent, string id, IStackProps props = nu
                     SourceArtifact = sourceArtifact,  // Where to get source code to build
                     CloudAssemblyArtifact = cloudAssemblyArtifact,  // Where to place built source
 
-                    InstallCommands = new [] 
-                    {
-                        "npm install -g aws-cdk", 
-                        "sudo apt-get install -y dotnet-sdk-3.1"
-                    },
-                    BuildCommands = new [] { "dotnet build" } // Language-specific build cmd
```
Comment (ryparker) [inline code/csharp/pipelines-workshop/src/CdkWorkshop/PipelineStack.cs:52]: Deprecated API params

File: workshop/content/40-dotnet/50-table-viewer/400-expose-table.md [hunk 8]
```diff
@@ -62,7 +62,7 @@ namespace CdkWorkshop
 
 ## Now we can access the table from our stack
 
-Go back to `CdkWorkshop.cs` and assign the `Table` property of the table viewer:
```
Comment (ryparker) [inline workshop/content/40-dotnet/50-table-viewer/400-expose-table.md:65]: CdkWorkshop is a directory not a file. This seems to be talking about `CdkWorkshopStack.cs`

**Review by NGL321 — APPROVED** [review]

**Pull request merged**